File: Sources/TePlannerKit/Capabilities/Tesla/Navigation.swift
import Foundation

/// `tesla.navigation.send` — push GPS coordinates to the car's nav.
/// Note: backend uses numeric vehicle_id (not VIN) for this one since
/// `navigation_gps_request` is one of the few endpoints not on the
/// VCP-signed path.
public struct SendNavigationCapability: Capability {
    public let id = "tesla.navigation.send"
    public let safetyClass: SafetyClass = .writable

    public init() {}

    public func invoke(
        ctx: CapabilityContext,
        params: [String: JSONValue],
        api: APIServiceProtocol
    ) async -> CapabilityResult {
        guard let lat = params.double("latitude"),
              let lng = params.double("longitude") else {
            return CapabilityResult(success: false, error: "latitude / longitude required")
        }
        let name = params.string("name") ?? ""
        guard let vehicleId = ctx.vehicleId else {
            return CapabilityResult(success: false, error: "vehicleId required")
        }
        let request = NavigationRequest(latitude: lat, longitude: lng, name: name)
        let result = await api.sendNavigation(vehicleId: vehicleId, request: request)
        switch result {
        case .success:
            return CapabilityResult(success: true)
        case .failure(let err):
            return CapabilityResult(success: false, error: err.localizedDescription)
        }
    }
}

File: backend/app/services/capabilities/tesla/navigation.py
"""Tesla navigation capabilities."""

from __future__ import annotations

from app.services.capabilities import register
from app.services.capabilities.base import (
    Capability,
    CapabilityCallContext,
    CapabilityResult,
    SafetyClass,
)


class SendNavigation(Capability):
    """Send GPS coordinates to vehicle nav. Uses numeric vehicle_id
    (not VIN) — navigation_gps_request is one of the few endpoints
    that doesn't go through the VCP signing path."""

    @property
    def id(self) -> str:
        return "tesla.navigation.send"

    @property
    def safety_class(self) -> SafetyClass:
        return SafetyClass.WRITABLE

    @property
    def params_schema(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "latitude": {"type": "number"},
                "longitude": {"type": "number"},
                "name": {"type": "string"},
                "order": {"type": "integer", "minimum": 1, "default": 1},
            },
            "required": ["latitude", "longitude"],
        }

    async def invoke(
        self, ctx: CapabilityCallContext, params: dict
    ) -> CapabilityResult:
        try:
            lat = float(params["latitude"])
            lon = float(params["longitude"])
        except (KeyError, ValueError, TypeError):
            return CapabilityResult(
                success=False, error="latitude / longitude required"
            )
        order = int(params.get("order", 1))
        await ctx.tesla_client.navigation_gps_request(
            vehicle_tag=ctx.vehicle_id,
            lat=lat,
            lon=lon,
            order=order,
        )
        return CapabilityResult(
            success=True,
            data={"destination": {"latitude": lat, "longitude": lon}},
        )


register(SendNavigation())
